internal/controller/multiclusterservice_controller.go

```diff
@@ -57,37 +67,71 @@ func (r *MultiClusterServiceReconciler) Reconcile(ctx context.Context, req ctrl.
 		return r.reconcileDelete(ctx, mcsvc)
 	}
 
-	if controllerutil.AddFinalizer(mcsvc, hmc.MultiClusterServiceFinalizer) {
-		if err := r.Client.Update(ctx, mcsvc); err != nil {
-			return ctrl.Result{}, fmt.Errorf("failed to update MultiClusterService %s with finalizer %s: %w", mcsvc.Name, hmc.MultiClusterServiceFinalizer, err)
+	return r.reconcile(ctx, mcsvc)
+}
+
+func (r *MultiClusterServiceReconciler) reconcile(ctx context.Context, mcs *hmc.MultiClusterService) (res ctrl.Result, err error) {
+	var servicesErr error
+	defer func() {
+		condition := metav1.Condition{
+			Reason: hmc.SucceededReason,
+			Status: metav1.ConditionTrue,
+			Type:   hmc.SveltosClusterProfileReadyCondition,
 		}
-		return ctrl.Result{}, nil
+
+		if err != nil {
+			condition.Message = err.Error()
+			condition.Reason = hmc.FailedReason
+			condition.Status = metav1.ConditionFalse
+		}
+
+		apimeta.SetStatusCondition(&mcs.Status.Conditions, condition)
+		err = errors.Join(err, servicesErr, r.updateStatus(ctx, mcs))
+	}()
+
+	if controllerutil.AddFinalizer(mcs, hmc.MultiClusterServiceFinalizer) {
+		if err := r.Client.Update(ctx, mcs); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update MultiClusterService %s with finalizer %s: %w", mcs.Name, hmc.MultiClusterServiceFinalizer, err)
+		}
+		return ctrl.Result{Requeue: true}, nil
```

Comment:
why do we need an explicit `requeue` here?

Reply:
The reason for such behavior is that adding a finalizer does not increase the generation of an object. So I would amend my initial question: do we need such a predicate for the main resource type being watched? I mean it's totally okay to have it, I just want to understand the reasoning behind this decision, since it might be tricky to debug. In general, this is a good decision to limit the number of events only to `spec` changes

Reply:
Without the `GenerationChangedPredicate{}`, the reconciliation went into a dead loop indefinitely.

Reply:
In an attempt to reproduce the dead loop:

- I removed the requeue:
```suggestion
		return ctrl.Result{}, nil
```

- And also removed the following:
```diff
func (r *MultiClusterServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
	return ctrl.NewControllerManagedBy(mgr).
-		For(&hmc.MultiClusterService{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		For(&hmc.MultiClusterService{}).
		Watches(&sveltosv1beta1.ClusterSummary{},
			handler.EnqueueRequestsFromMapFunc(requeueSveltosProfileForClusterSummary),
			builder.WithPredicates(predicate.Funcs{
				DeleteFunc:  func(event.DeleteEvent) bool { return false },
				GenericFunc: func(event.GenericEvent) bool { return false },
			}),
		).
		Complete(r)
}
```

- Then I created a ManagedCluster with `servicesPriority=100`:
- Then I created a MultiClusterService with `servicesPriority=50` but did not observe the dead loop.
- So I changed the `servicesPriority=1000` for the MultiClusterService but still did not observe the dead loop as indicated by `generation: 2` below:
```sh
➜  hmc git:(services-status) khmc get multiclusterservices.hmc.mirantis.com global-ingress -o yaml
apiVersion: hmc.mirantis.com/v1alpha1
kind: MultiClusterService
metadata:
  . . .
  finalizers:
  - hmc.mirantis.com/multicluster-service
  generation: 2
  name: global-ingress
  resourceVersion: "10082"
  uid: 181fe130-95a6-493f-a9e4-dd3fcff009dc
spec:
  clusterSelector:
    matchLabels:
      app.kubernetes.io/managed-by: Helm
  services:
  - name: ingress-nginx
    namespace: ingress-nginx
    template: ingress-nginx-4-11-3
  servicesPriority: 1000
  stopOnConflict: false
status:
  conditions:
  - lastTransitionTime: "2024-10-31T07:05:35Z"
    message: ""
    reason: Succeeded
    status: "True"
    type: SveltosClusterProfileReady
  - lastTransitionTime: "2024-10-31T07:05:35Z"
    message: ""
    reason: Succeeded
    status: "True"
    type: FetchServicesStatusSuccess
  - lastTransitionTime: "2024-10-31T07:05:35Z"
    message: MultiClusterService is ready
    reason: Succeeded
    status: "True"
    type: Ready
  observedGeneration: 2
  services:
  - clusterName: wali-aws-dev
    clusterNamespace: hmc-system
    conditions:
    - lastTransitionTime: "2024-10-31T07:06:52Z"
      message: ""
      reason: Provisioned
      status: "True"
      type: Helm
    - lastTransitionTime: "2024-10-31T07:06:32Z"
      message: Release ingress-nginx/ingress-nginx
      reason: Managing
      status: "True"
      type: ingress-nginx.ingress-nginx/SveltosHelmReleaseReady
```

So this would suggest that there is no bug in the Idempotence property of the reconcile. I think I may have experienced it while I was still in the process of implementing status updates in reconcile. 

Nevertheless, I think we should keep the `GenerationChangedPredicate{}` and I added a comment explaining why the requeue is needed.

Reply:
Ok, thanks for checking it on-hands. I'm also okay with the predicate